c2rust-transpile/src/c_ast/mod.rs

```diff
@@ -2183,6 +2183,177 @@ impl CTypeKind {
             _ => unimplemented!("Printer::print_type({:?})", self),
         }
     }
+
+    /// Whether `value` is guaranteed to be in this integer type's range.
+    /// Thus, minimum ranges are used (e.x. [`i16::MIN`] for [`Self::Int`]).
+    pub fn guaranteed_integer_in_range(&self, value: u64) -> bool {
+        fn in_range<T: TryFrom<u64>>(value: u64) -> bool {
+            T::try_from(value).is_ok()
+        }
+
+        use CTypeKind::*;
+        match *self {
+            Void => false,
+
+            Bool => (0..=1).contains(&value),
+
+            // Can be signed or unsigned, so choose the minimum range of each.
+            Char => (u8::MIN as u64..=i8::MAX as u64).contains(&value),
+            WChar => in_range::<i32>(value),
+
+            // `int` is at least `i16` and `long` is at least `i32`.
+            SChar => in_range::<i8>(value),
+            Short => in_range::<i16>(value),
+            Int => in_range::<i16>(value),
+            Long => in_range::<i32>(value),
+            LongLong => in_range::<i64>(value),
+
+            // `unsigned int` is at least `u16` and `unsigned long` is at least `u32`.
+            UChar => in_range::<u8>(value),
+            UShort => in_range::<u16>(value),
+            UInt => in_range::<u16>(value),
+            ULong => in_range::<u32>(value),
+            ULongLong => in_range::<u64>(value),
+
+            Int8 => in_range::<i8>(value),
+            Int16 => in_range::<i16>(value),
+            Int32 => in_range::<i32>(value),
+            Int64 => in_range::<i64>(value),
+            Int128 => in_range::<i128>(value),
+
+            UInt8 => in_range::<u8>(value),
+            UInt16 => in_range::<u16>(value),
+            UInt32 => in_range::<u32>(value),
+            UInt64 => in_range::<u64>(value),
+            UInt128 => in_range::<u128>(value),
+
+            // There's no guarantee on pointer size, but `NULL` should work.
+            IntPtr => value == 0,
+            UIntPtr => value == 0,
+
+            IntMax => in_range::<i64>(value),
+            UIntMax => in_range::<u64>(value),
+
+            // `size_t` is at least a `u16`, and similar for `ssize_t` and `ptrdiff_t`.
+            Size => in_range::<u16>(value),
+            SSize => in_range::<i16>(value),
+            PtrDiff => in_range::<i16>(value),
+
+            // Floats, see `Self::guaranteed_float_in_range`.
+            Float => false,
```

Comment:
Technically this still works if the integer fits into the mantissa, i.e., 24 bits.